File: go/problem_52.go
package main

import "container/list"

type LRUCache struct {
	cap   int
	items map[int]*list.Element
	list  *list.List
}

type lruEntry struct{ key, val int }

func newLRUCache(cap int) *LRUCache {
	return &LRUCache{cap: cap, items: map[int]*list.Element{}, list: list.New()}
}

func (c *LRUCache) get(key int) (int, bool) {
	if el, ok := c.items[key]; ok {
		c.list.MoveToFront(el)
		return el.Value.(*lruEntry).val, true
	}
	return 0, false
}

func (c *LRUCache) set(key, val int) {
	if el, ok := c.items[key]; ok {
		el.Value.(*lruEntry).val = val
		c.list.MoveToFront(el)
		return
	}
	if c.list.Len() == c.cap {
		back := c.list.Back()
		c.list.Remove(back)
		delete(c.items, back.Value.(*lruEntry).key)
	}
	el := c.list.PushFront(&lruEntry{key, val})
	c.items[key] = el
}

func problem52() {
	println("Problem 52:")
	c := newLRUCache(2)
	c.set(1, 1)
	c.set(2, 2)
	v, ok := c.get(1)
	if !ok || v != 1 {
		panic("assertion failed")
	}
	c.set(3, 3)
	_, ok = c.get(2)
	if ok {
		panic("2 should have been evicted")
	}
	v, ok = c.get(3)
	if !ok || v != 3 {
		panic("assertion failed")
	}
	println("All Tests Passed")
}

File: typescript/problem_52.ts
export class LRUCache52 {
    private cap: number;
    private cache: Map<number, number>;
    constructor(cap: number) { this.cap = cap; this.cache = new Map(); }
    get(key: number): number | null {
        if (!this.cache.has(key)) return null;
        const v = this.cache.get(key)!;
        this.cache.delete(key); this.cache.set(key, v);
        return v;
    }
    set(key: number, val: number): void {
        if (this.cache.has(key)) this.cache.delete(key);
        this.cache.set(key, val);
        if (this.cache.size > this.cap) this.cache.delete(this.cache.keys().next().value!);
    }
}

export function problem52(): void {
    console.log("Problem 52:");
    const c = new LRUCache52(2);
    c.set(1,1); c.set(2,2);
    if (c.get(1) !== 1) throw new Error("assertion failed");
    c.set(3,3);
    if (c.get(2) !== null) throw new Error("2 should be evicted");
    if (c.get(3) !== 3)  throw new Error("assertion failed");
    console.log("All Tests Passed");
}
